File: test/python2/test_utils.py
from datetime import date, timedelta

from erlport import Atom, erlang

def switch(n):
    result = 0
    for i in range(n):
        _, result = erlang.call(Atom("python2_tests"), Atom("test_callback"),
            [(result, i)])
    return n

def setup_message_handler():
    def handler(message):
        erlang.call(Atom("python2_tests"), Atom("test_callback"),
            [(Atom("message"), message)])
    erlang.set_message_handler(handler)
    return Atom("ok")

def setup_faulty_message_handler():
    def handler(message):
        raise ValueError(message)
    erlang.set_message_handler(handler)
    return Atom("ok")

def recurse(python, n):
    if n <= 0:
        return Atom("done")
    return erlang.call(Atom("python2_tests"), Atom("recurse"),
        [python, n - 1])

def identity(v):
    return v

def length(v):
    return len(v)

def print_string(s):
    print s.to_string()

class TestClass(object):
    class TestSubClass(object):
        @staticmethod
        def test_method():
            return Atom("ok")

def setup_date_types():
    erlang.set_encoder(date_encoder)
    erlang.set_decoder(date_decoder)
    return Atom("ok")

def date_encoder(value):
    if isinstance(value, date):
        value = Atom("date"), (value.year, value.month, value.day)
    elif isinstance(value, timedelta):
        value = Atom("days"), value.days
    return value

def date_decoder(value):
    if isinstance(value, tuple) and len(value) == 2:
        if value[0] == "date":
            year, month, day = value[1]
            value = date(year, month, day)
        elif value[0] == "days":
            value = timedelta(days=value[1])
    return value

File: test/ruby1.9/test_utils.rb
include ErlPort
include ErlPort::ErlTerm

def switch n
    result = 0
    for i in 0...n
        _, result = Erlang.call :ruby19_tests, :test_callback, \
            [Tuple.new([result, i])]
    end
    n
end

def setup_message_handler
    Erlang.set_message_handler {|message|
        Erlang.call :ruby19_tests, :test_callback, \
            [Tuple.new([:message, message])]
    }
    :ok
end

def setup_faulty_message_handler
    Erlang.set_message_handler {|message|
        raise ValueError, message
    }
    :ok
end

def recurse ruby, n
    if n <= 0
        :done
    else
        Erlang.call :ruby19_tests, :recurse, [ruby, n - 1]
    end
end

def identity v
    v
end

def add a, b
    a + b
end

def len s
    s.length
end

def print_string s
    puts s.pack "U*"
end

def string_to_sym s
    s.pack("U*").to_sym
end

module TestModule
    class TestClass
        def self.test_method
            :ok
        end
    end
end

def setup_date_types
    Erlang.set_encoder {|v| date_encoder v}
    Erlang.set_decoder {|v| date_decoder v}
    :ok
end

def date_encoder value
    if value.is_a? Time
        value = Tuple.new([:date,
            Tuple.new([value.year, value.month, value.day])])
    end
    value
end

def date_decoder value
    if value.is_a? Tuple and value.length == 2 and value[0] == :date
        year, month, day = value[1]
        value = Time.utc(year, month, day)
    end
    value
end
